### PR TITLE
Insgradebook

### PR DESCRIPTION
**Pull request by dtracers**
This contains the very core basic functionality of a gradebook

you can add a grade and a comment
and you can view a grade.

This does nothing more than that.

Depends on #482, #483

### CODE REVIEW

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/GradeManager.java [hunk 1]
```diff
@@ -362,11 +362,15 @@ static ProtoGrade getGrade(final Authenticator authenticator, final DB dbs, fina
                 .append(COURSE_PROBLEM_ID, new BasicDBObject(EXISTS, false));
         final BasicDBObject sortMethod = new BasicDBObject(ASSIGNMENT_ID, 1).append(USER_ID, 1); // Sort by assignmentId then userId
         final DBCursor cursor = gradeCollection.find(query).sort(sortMethod);
+
+        final List<ProtoGrade> grades = new ArrayList<>();
+
         if (!cursor.hasNext()) {
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/database/institution/mongo/GradeManager.java]: It says dont let this merge with master lol

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java [hunk 2]
```diff
@@ -6,6 +6,9 @@
 import database.institution.Institution;
 import protobuf.srl.grading.Grading;
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java]: add a new line here plz

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java [hunk 3]
```diff
@@ -25,13 +28,23 @@ private GradingUpsertHandler() {
...
      * @throws DatabaseAccessException Thrown if there is something not found in the database.
      * @throws InvalidProtocolBufferException Thrown if a protobuf object is not correctly formatted.
      */
-    public static void gradingUpsertHandler(final Institution institution, final Data.ItemSend itemSet, final String authId)
+    public static void gradingUpsertHandler(final Institution institution, final Data.ItemSend itemSet, final String authId, final long gradedTime)
             throws AuthenticationException, DatabaseAccessException, InvalidProtocolBufferException {
         final Grading.ProtoGrade grade = Grading.ProtoGrade.parseFrom(itemSet.getData());
-        institution.addGrade(authId, grade);
+        final Grading.ProtoGrade.Builder clone =  Grading.ProtoGrade.newBuilder(grade);
+        final List<Grading.GradeHistory.Builder> gradeHistroy = clone.getGradeHistoryBuilderList();
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java]: misspelled History, you wrote: "Histroy"

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java [hunk 4]
```diff
@@ -25,13 +28,23 @@ private GradingUpsertHandler() {
...
+        if (gradeHistroy.size() == 1) {
+            final Grading.GradeHistory.Builder newestHistroy = gradeHistroy.get(0);
```
Comment (areftw) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java]: How did this compile if "gradeHistory" is misspelled?! Please run maven

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.css [hunk 5]
```diff
@@ -1,22 +1,93 @@
...
+
+.gradeSelected {
+    padding-top:2px;
+}
+
+.gradeForm {
+    margin: 0px;
+}
+
+/*
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.css]: Is this all supposed to be commented??

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.css [hunk 6]
```diff
@@ -25,8 +96,8 @@ th {
     background-color: #C2E0FF;
 }
 
-.tabletalk: hover td  div {
-    background-color: blue;
+.tabletalk: hoverable td  div {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.css]: Its all commented out

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js [hunk 7]
```diff
@@ -1,25 +1,479 @@
...
+            CourseSketch.gradeBook.loadGrades(courseId);
+            CourseSketch.gradeBook.createTabs([ 'Quiz', 'Homework', 'Test' ], document.querySelector('.tabholder'));
+        });
+        var table = $('.tabletalk');
+        CourseSketch.gradeBook.initializeTableScrolling(document.querySelector('.tabletalk'), table.offset());
+        table.keyup(keyEventHandler);
+    });
+
+    /**
+     * Loads grades for the given course.
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: Need a new line after this line

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js [hunk 8]
```diff
@@ -1,25 +1,479 @@
...
+     */
+    CourseSketch.gradeBook.initializeTable = function(assignmentList, gradeList, idToNameMap, table) {
+        table.innerHTML = '';
+        var studentList = Array.from(idToNameMap.keys());
+        var assignmentMap = new Map();
+        var idToRowMap = new Map();
+        var body = document.createElement('tbody');
+        table.appendChild(body);
+
+        for (var i = 0; i < studentList.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: use a for in loop here please like:
`for (student in studentList) {..}`

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js [hunk 9]
```diff
@@ -1,25 +1,479 @@
...
+    /**
+     * Creates a new tr with the number of assignments and adds it to the student map.
+     *
+     * @param {Map<String,Element>} idToRowMap This is a map of studentId's to table rows.
+     * @param {String} studentId The student's ID number.
+     * @param {Integer} assignmentList The list of assignments for the course.
+     */
+    function addNewStudent(idToRowMap, studentId, assignmentList, table) {
+        var row = document.createElement('tr');
+        for (var i = 0; i < assignmentList.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: for in loop here too

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js [hunk 10]
```diff
@@ -1,25 +1,479 @@
...
+     * @param {List<ProtoGrade>} listGrades The list of grades from the server.
+     * @param {Map<String, Integer>} idToRowMap This is a map of studentIds to table rows.
+     * @param {Map<String, Integer>} assignmentMap This is a map of assignmentIds to table columns.
+     * @param {HTMLTable} table The grade table on the webpage.
+     * @return {List<ProtoGrade>} grades that were not displayed.
+     *          This is because the users do not exist anymore in the course roster.
+     */
+    CourseSketch.gradeBook.populateGrades = function(listGrades, idToRowMap, assignmentMap, table) {
+        var gradesNotShown = [];
+        for (var i = 0; i < listGrades.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: for in loop here too

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js [hunk 11]
```diff
@@ -1,25 +1,479 @@
...
+            });
+            */
+            button.className = 'waves-effect waves-teal btn-flat truncate';
+            th.appendChild(button);
+            row.appendChild(th);
+        }
+
+        // Remove this code once the single request method has been resolved.
+        CourseSketch.dataManager.getAssignments(assignmentList, undefined, function(assignments) {
+            for (var i = 0; i < assignments.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: for in loop here

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js [hunk 12]
```diff
@@ -1,25 +1,479 @@
...
+     * @param {List<String>} gradeCategories The list of categories for grades in the gradebook.
+     * @param {element} tabholder The element that will hold the tabs.
+     */
+    CourseSketch.gradeBook.createTabs = function(gradeCategories, tabholder) {
+        tabholder.innerHTML = '';
+        var tabs = document.createElement('ul');
+        tabs.className = 'tabs';
+        gradeCategories.unshift('All');
+        gradeCategories.push('New');
+        for (var i = 0; i < gradeCategories.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: for in loop here

File: CourseSketchProjects/coursesketchwebclient/src/main/src/utilities/persistantData/specificManager/courseDataManager.js [hunk 13]
```diff
@@ -390,8 +398,14 @@ function CourseDataManager(parent, advanceDataListener, database, Request, ByteB
                 return;
             }
 
-            var decodedRoster = CourseSketch.PROTOBUF_UTIL.decodeProtobuf(item.data[0], CourseSketch.PROTOBUF_UTIL.getIdChainClass());
-            callback(decodedRoster.idChain);
+
+            var decodedRoster = CourseSketch.PROTOBUF_UTIL.decodeProtobuf(item.data[0], CourseSketch.prutil.getUserNameResponseClass());
+            var userMap = new Map();
+            for (var i = 0; i < decodedRoster.userNames.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/utilities/persistantData/specificManager/courseDataManager.js]: use a for in loop here too

File: CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html [hunk 14]
```diff
@@ -10,11 +10,99 @@
...
+		var assignmentList = [];
+		for (var i = 0; i < 10; i++) {
+			assignmentList.push('' + (i + 11));
+		}
+
+        var studentList = ["student1", "stud2", "gilbert", "gabby", "student5", "student6",
+        "matt", "stud8", "stud9", "stud10"];
+
+        var idToNameMap = new Map();
+        for (var i = 0; i < studentList.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html]: for in loop

File: CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html [hunk 15]
```diff
@@ -10,11 +10,99 @@
...
+            gradeList.push(grade);
+
+            var grade = CourseSketch.PROTOBUF_UTIL.ProtoGrade();
+            grade.courseId = courseId;
+            grade.userId = studentList[1];
+            grade.assignmentId = assignmentList[2];
+            grade.currentGrade = 95;
+            gradeList.push(grade);
+
+            for (var i = 4; i < studentList.length; i++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html]: for in loop

File: CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html [hunk 16]
```diff
@@ -10,11 +10,99 @@
...
+                for (var j = 0; j < assignmentList.length; j++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html]: for in loop

File: CourseSketchProjects/coursesketchwebclient/src/test/src/testUtilities/fakePage/fakeTestData/courseProblemData.js [hunk 17]
```diff
@@ -225,4 +225,20 @@
     bankProblem14.id = problem14.id;
     CourseSketch.fakeProblems.push(problem14);
     CourseSketch.fakeBankProblems.push(bankProblem14);
+	
+	for (var i = 0; i < 10; i++) {
+		for (var j = 0; j < 10; j++) {
+	       var problemLoop = CourseSketch.PROTOBUF_UTIL.SrlProblem();
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/testUtilities/fakePage/fakeTestData/courseProblemData.js]: tab this please

File: CourseSketchProjects/coursesketchwebclient/src/test/src/testUtilities/fakePage/fakeTestData/courseProblemData.js [hunk 18]
```diff
@@ -225,4 +225,20 @@
     bankProblem14.id = problem14.id;
     CourseSketch.fakeProblems.push(problem14);
     CourseSketch.fakeBankProblems.push(bankProblem14);
+	
+	for (var i = 0; i < 10; i++) {
+		for (var j = 0; j < 10; j++) {
```
Comment (areftw) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/testUtilities/fakePage/fakeTestData/courseProblemData.js]: this is not tabbed correctly

File: CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java (hunk 2 above)
Comment (dtracers) [inline CourseSketchProjects/CourseSketchDatabaseServer/src/main/java/handlers/subhandlers/GradingUpsertHandler.java]: what?

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js (hunk 8 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: that does not work in javascript.  (sorry)

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js (hunk 9 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: still does not work in this language

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js (hunk 10 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: not possible

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js (hunk 11 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: not possible again

File: CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js (hunk 12 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/instructor/gradebook/instructorGradebook.js]: still not possible

File: CourseSketchProjects/coursesketchwebclient/src/main/src/utilities/persistantData/specificManager/courseDataManager.js (hunk 13 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/main/src/utilities/persistantData/specificManager/courseDataManager.js]: still not possible.

File: CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html (hunk 14 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html]: ill let you know when this is possible

File: CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html (hunk 15 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html]: not possible

File: CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html (hunk 16 above)
Comment (dtracers) [inline CourseSketchProjects/coursesketchwebclient/src/test/src/instructor/gradebook/instructorGradebookFakePage.html]: not possible

**Pull request merged**